File: hash_table/3/my.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if not s:
            return 0
        
        char_map = {}  # Maps character to its most recent position
        start = 0      # Start of current substring
        max_length = 0
        
        for end, char in enumerate(s):
            # If we see a duplicate character within our current window
            if char in char_map and char_map[char] >= start:
                # Move the start pointer to position after the earlier occurrence
                start = char_map[char] + 1
            # Update the max length
            max_length = max(max_length, end - start + 1)
            # Record this character's position
            char_map[char] = end
        
        return max_length
        

File: hash_table/3/my.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        
        int maxLength = 0;
        int start = 0;
        int[] charIndex = new int[128]; // ASCII characters
        
        // Initialize the array with -1 to indicate character hasn't been seen
        for (int i = 0; i < 128; i++) {
            charIndex[i] = -1;
        }
        
        for (int end = 0; end < s.length(); end++) {
            char c = s.charAt(end);
            
            // If this character was seen and its index is >= our current window start
            if (charIndex[c] >= start) {
                // Move start to position after the previous occurrence
                start = charIndex[c] + 1;
            }
            
            // Update maximum length
            maxLength = Math.max(maxLength, end - start + 1);
            
            // Record this character's position
            charIndex[c] = end;
        }
        
        return maxLength;
    }
}